WIP: DRA kubelet connection monitoring

bart0sh opened the pull request:
Please, ignore.

This is a fork of https://github.com/kubernetes/kubernetes/pull/132058 to implement more e2e tests. The result will be merged back to the original PR.

bart0sh committed:
DRA kubelet: add connection monitoring

This ensures that ResourceSlices get removed also when a plugin becomes
unresponsive without removing the registration socket.

Tests are from https://github.com/kubernetes/kubernetes/pull/131073 by Ed
with some modifications, the implementation is new.

pohly committed:
DRA kubelet: simplify plugin creation and connection handling

Instead of creating the gRPC connection on demand and forcing gRPC to connect,
we establish it immediately and rely on gRPC to handle the underlying
connection automatically like it usually does.

It's not clear what benefit the one second connection timeout had. The way it
is now, gRPC calls still fail when the underlying connection cannot be
established. Having to have a separate context for establishing that connection
just made the code more complex.

The DRAPluginManager is the central component which manages plugins. Making it responsible
for creating them reduces the number of places where a DRAPlugin struct needs to
be initialized. Doing this in the DRAPluginManager instead of a stand-alone function
simplifies the implementation of connection monitoring, because that will be
something that is tied to the DRAPluginManager state.

pohly committed:
DRA kubelet: use TimedWorkersQueue

Conceptually TimedWorkersQueue is similar to the current code: it spawns
goroutines and cancels them. Using it makes the code a bit shorter, even though
the TimedWorkersQueue API could be a bit nicer and more consistent (key string
vs. WorkerArgs as parameters).

Depending on the tainteviction package is a bit odd, which is the reason why
TimedWorkersQueue wasn't already used earlier. But there don't seem to be other
implementations of this common
problem. https://pkg.go.dev/k8s.io/client-go/util/workqueue#TypedDelayingInterface
doesn't work because queue entries cannot be removed.

This doesn't really solve the problem of tracking goroutines for wiping because
TimedWorkersQueue doesn't support that. But not tracking is arguably better
than doing it wrong and this only affects unit tests, so it should be okay.

bart0sh commented:
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh committed:
kubelet: DRA: fix linter warnings

Fixed the following warnings:
dra_test.go:884:2: singleCaseSwitch: should rewrite switch statement to if statement (gocritic)
	switch podName {
	^
dra_test.go:686:4: SA4006: this value of kubeletPlugin is never used (staticcheck)
	kubeletPlugin = newDRAService(ctx, f.ClientSet, nodeName, driverName)
        ^

bart0sh commented:
/test pull-kubernetes-linter-hints

bart0sh commented:
/test pull-kubernetes-linter-hints
/test pull-kubernetes-unit-windows-master
/test pull-kubernetes-kind-dra-n-1
/test pull-kubernetes-kind-dra-n-2

bart0sh commented:
/test pull-kubernetes-linter-hints
/test pull-kubernetes-unit-windows-master
/test pull-kubernetes-kind-dra-n-1
/test pull-kubernetes-kind-dra-n-2
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh commented:
/test pull-kubernetes-linter-hints
/test pull-kubernetes-unit-windows-master
/test pull-kubernetes-kind-dra-n-1
/test pull-kubernetes-kind-dra-n-2
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh commented:
/test pull-kubernetes-unit-windows-master

bart0sh commented:
/test pull-kubernetes-linter-hints
/test pull-kubernetes-unit-windows-master
/test pull-kubernetes-kind-dra-n-1
/test pull-kubernetes-kind-dra-n-2
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh committed:
e2e_node: DRA: pass socket path as a parameter

Added an ability to specify the socket path for the DRA gRPC
service in the e2e node tests.

The PluginSocket option is added to allow setting the name
of the socket inside the directory where the DRA driver
creates the socket for the DRA gRPC calls. This is used by
the kubelet to connect to the DRA plugin.

The newDRAService and newRegistrar functions are updated to
accept a socketPath parameter, which is used to configure
the PluginDataDirectoryPath and PluginSocket options for the
DRA plugin.

This change enables more flexible configuration of the DRA
plugin in e2e tests, allowing for testing with different
socket paths.

bart0sh committed:
e2e_node: DRA: add tests for different socket setups

Added tests to verify DRA functionality with 2 different socket
configurations:
- the same socket is used for the registration and the DRA service
- 2 separate sockets are used for the registration and the DRA service

Used table-driven ginkgo to avoid code duplication:
specs https://onsi.github.io/ginkgo/#table-driven-tests

This change enhances the robustness of the DRA e2e tests by
validating its behavior with different socket setups.

bart0sh commented:
/test pull-kubernetes-linter-hints
/test pull-kubernetes-unit-windows-master
/test pull-kubernetes-kind-dra-n-1
/test pull-kubernetes-kind-dra-n-2
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh commented:
@pohly This PR is ready for your review. PTAL.

pohly commented on test/e2e_node/dra_test.go:
```diff
@@ -683,7 +683,7 @@ var _ = framework.SIGDescribe("node")(framework.WithLabel("DRA"), feature.Dynami
 
 			ginkgo.By("restarting plugin")
-			kubeletPlugin = newDRAService(ctx, f.ClientSet, nodeName, driverName)
+			newDRAService(ctx, f.ClientSet, nodeName, driverName)
```
I intentionally updated `kubeletPlugin` here, ignoring the linter. If we ever add new code below which then starts to use it, that code would use a stale value.

pohly replied:
Hints are not failures. Developers should use their own judgment to decide whether the reported issues should be addressed.

pohly commented on test/e2e_node/dra_test.go:
```diff
@@ -881,8 +881,7 @@ func createTestObjects(ctx context.Context, clientSet kubernetes.Interface, node
 	}
-	switch podName {
-	case "drasleeppod":
+	if podName == "drasleeppod" {
```
No strong opinion about this one.

bart0sh replied:
The same thing. I wouldn't touch it if CI is green.

pohly commented on pkg/kubelet/cm/dra/plugin/dra_plugin_test.go:
```diff
@@ -59,6 +60,47 @@ func (f *fakeGRPCServer) NodeUnprepareResources(ctx context.Context, in *drapbv1
 // tearDown is an idempotent cleanup function.
 type tearDown func()
 
+// createTempSocketPath creates a temporary directory for use in tests and
```
This commit should go into a separate PR.

bart0sh replied:
Then we should probably shorten test/subtests or socket name for the `TestConnectionHandling` test or ignore `pull-kubernetes-unit-windows-master` failure in the original branch. Or wait when that PR is merged, which I'd rather avoid.

pohly replied:
I intend to ignore `pull-kubernetes-unit-windows-master` because we know that the failure is unrelated to the PR.

pohly replied:
Or is it passing on master and only failing in the new tests?

bart0sh replied:
Yes, it is. It fails only on the `TestConnectionHandling`. It's mentioned in the commit message:
```
listen unix
C:\Users\azureuser\AppData\Local\Temp\TestConnectionHandlingno-wipe-slices-on-reconnect3094073874\001/dra-plugin-test.sock:
bind: invalid argument
```

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -261,6 +261,16 @@ func PluginDataDirectoryPath(path string) Option {
 
+// PluginSocket sets the name of the socket inside the directory where
+// the DRA driver creates the socket for the DRA gRPC calls. This is used
+// by the kubelet to connect to the DRA plugin.
```
Add:
```
//
// This is meant for testing. Normal DRA drivers should
// not use this and instead rely on the automatic handling
// of the name.
```

bart0sh commented:
/retest

bart0sh committed:
kubelet: DRA: fix test failure on Windows

Unix socket path has to be <= 108 characters in length on Windows.
Shortened DRA socket path for TestConnectionHandling and
TestRegistrationHandler tests should fix the test run on Windows.

bart0sh commented:
/test pull-kubernetes-linter-hints
/test pull-kubernetes-unit-windows-master

bart0sh committed:
kubelet: DRA: fix unit test failure

Fixed race condition caused by removing already removed
socket directory.

bart0sh committed:
kubelet: DRA: fix TestRegistrationHandler

Using the same socket path for different test cases caused
test failure on windows:
listen unix C:\Users\azureuser\AppData\Local\Temp\TestRegistrationHandler3881105518\001/dra-plugin-a.sock:
bind: Only one usage of each socket address (protocol/network address/port) is normally permitted.

Creating unique socket path for every test case should fix it.

pohly reviewed (commented):
Some nits. I'll copy your commits into my branch and then change them there, no need to update your branch.

pohly commented:
I updated https://github.com/kubernetes/kubernetes/pull/132058, with some changes to your commits to address what I called out above.

pohly commented:
/close

Code moved to https://github.com/kubernetes/kubernetes/pull/132058
